File: modules/api/src/main/java/com/natevaughan/kchat/api/Message.java
package com.natevaughan.kchat.api;

import java.time.Instant;

/**
 * Created by nate on 3/8/19
 */
public class Message {

	private final Long id;
	private final Instant dateCreated;
	private final String text;
	private final User author;
	private final Chat chat;
	private final Instant lastEdited;

	public Message(Long id, Instant dateCreated, String text, User author, Chat chat, Instant lastEdited) {
		this.id = id;
		this.dateCreated = dateCreated;
		this.text = text;
		this.author = author;
		this.chat = chat;
		this.lastEdited = lastEdited;
	}

	public Message(String text, User author, Chat chat) {
		this.id = null;
		this.dateCreated = null;
		this.text = text;
		this.author = author;
		this.chat = chat;
		this.lastEdited = null;
	}

	public Long getId() {
		return id;
	}

	public Instant getDateCreated() {
		return dateCreated;
	}

	public String getText() {
		return text;
	}
	public User getAuthor() {
		return author;
	}

	public Chat getChat() {
		return chat;
	}

	public Instant getLastEdited() {
		return lastEdited;
	}

	public Message copy(String text) {
		return new Message(id, dateCreated, text, author, chat, Instant.now());
	};
}

File: modules/chat-api/src/main/kotlin/com/natevaughan/kchat/api/Message.kt
package com.natevaughan.kchat.api

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty

import java.time.Instant

/**
 * Created by nate on 3/8/19
 */
class Message(
		val id: String,
		@get:JsonIgnore val dateCreated: Instant,
		val text: String,
		val author: User,
		@get:JsonIgnore val chat: Chat,
		@get:JsonIgnore private val lastEdited: Instant?) {

	@JsonProperty("lastEdited")
	fun lastEditedJson(): Long? {
		return lastEdited?.toEpochMilli()
	}

	@JsonProperty("dateCreated")
	fun dateCreatedJson(): Long? {
		return lastEdited?.toEpochMilli()
	}

	fun copy(text: String): Message {
		return Message(id, dateCreated, text, author, chat, Instant.now())
	}
}
